pkg/binary/envoytest/util.go

```diff
@@ -64,19 +39,21 @@ func RequireRunTerminate(t *testing.T, terminate func(r *envoy.Runtime), r *envo
 		}
 	}
 
+	// tee the error stream so we can look for the "started" line without consuming it.
+	stderr := new(bytes.Buffer)
+	r.Err = io.MultiWriter(r.Err, stderr)
+
 	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
-		if err := r.Run(ctx, args); err != nil {
-			t.Errorf("unable to run %v: %v", r, err)
-			return
-		}
+		err = r.Run(ctx, args)
 		cancel()
 	}()
 
+	reader := bufio.NewReader(stderr)
 	require.Eventually(t, func() bool {
-		_, err := r.GetPid()
-		return err == nil
-	}, 1*time.Second, 100*time.Millisecond, "never started process")
+		b, e := reader.Peek(512)
+		return e != nil && strings.Contains(string(b), "started\n")
```

Comment:
more precise and avoids flakes. If we exit as soon as we see a pid, we could kill the process slightly before traps hook. Yes, this took me a long time to figure out :D

Reply:
good catch and you saved millions of people 😄